components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java

```diff
@@ -570,7 +571,7 @@ private void sendResponse(HttpServletRequest req, HttpServletResponse resp, Stri
             }
 
             PrintWriter out = resp.getWriter();
-            out.print(finalPage);
+            out.print(Encode.forHtml(finalPage));
```

Comment:
It's wrong to encode the full html page which is constructed via line 554-570 here. If you encode so, none of the HTML markups will render. Instead encode untrusted parts such as response, relayState ( relayState is URL encoded which is wrong in this context) etc. Further, make sure they are not re-encoded